File: Homework/Ngan_Luu/141-linked-list-cycle.py
#https://leetcode.com/problems/linked-list-cycle/
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        p1 = head
        p2 = head
        while p2 is not None and p2.next is not None:
            p2 = p2.next.next
            p1 = p1.next
            if p1 == p2:
                return True
        return False

File: Homework/Son/W1/141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
 var hasCycle = function(head) {
    if (head === null) return false
    let slow = head;
    let fast = head;
    while (fast !== null) {
        if (fast.next !== null)  {
            fast = fast.next; slow = slow.next;
        } else return false;
        if (fast.next !== null) {fast = fast.next} 
        else return false;
        if (fast == slow) return true
    }
};

// https://leetcode.com/problems/linked-list-cycle/